Hotel 6_1_

dudarenko-dmitry opened the pull request:
Hi, Alex.
I need your help :)
I added to my application ability to import/export of all entities. But there are some difficulties in import of Orders (in class ReaderOrder).
I can't understand why parser doesn't read the last field. 
I checked names of field in Writer, Reader, Parser (everything ok), changed several times or used default Separators, QuoterChar, EscapeChar but get everytime the same results.
![image](https://github.com/dudarenko-dmitry/senla_training/assets/102149674/564cc867-4524-46b5-8774-b2268560eb21)

field in file is not emtpy. I save here List<Interger> which was previously converted to String.
![image](https://github.com/dudarenko-dmitry/senla_training/assets/102149674/9cbac4ac-ea81-4167-84c6-bdbae1caae73)

dudarenko-dmitry committed:
TASK_6.
New branch Hotel_6 was created.
README.md was updated.

dudarenko-dmitry committed:
TASK_6.
Renamed all "unreadable" classes in UserInterface
Create in UI menu items for save/load Application's Entities.
Made all entities implement Serializable.
Add Reader and Writer for all entities.
Change int to Integer in id-fields of Entities.
In Writers added null safe checking.
Changed String to Enum in all Facilities' and HotelServices' entities.

dudarenko-dmitry committed:
TASK_6.
Create in UI menu items for save/load Application's Entities.
Add Reader and Writer for entity Room.

dudarenko-dmitry committed:
TASK_6.
Created Reader and Writer for entities Guest, HotelServices(RoomReservation), Order.
Made all entities implement Serializable.

Mark005 commented on src/pl/senla/hotel/entity/CsvOrder.java:
```diff
@@ -0,0 +1,60 @@
+
+public class CsvOrder implements Serializable {
+
+    private int idOrder = -1;
```
don't use primitives, use wrappers instead and you will avoid magical numbers as -1

dudarenko-dmitry replied:
changed

dudarenko-dmitry replied:
changed

Mark005 commented on src/pl/senla/hotel/ie/ReaderOrder.java:
```diff
@@ -0,0 +1,54 @@
+        CsvToBean<CsvOrder> orderCsvToBean = new CsvToBean<>();
+        ICSVParser parser = new CSVParserBuilder()
+                .withSeparator(';')
+                .withQuoteChar('/')
```
i think a bug could be here, if I'm wrong then try to debug it on your own because it's your responsibility as a developer to solve problems with a code and provide a working version of your code.

dudarenko-dmitry replied:
I haven't found mistakes in these lines.
Changed all logic in Class and use BufferedReader with additional converter which create Order from received String. Such code works.

Mark005 commented on src/pl/senla/hotel/ie/ReaderOrder.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class ReaderOrder implements Reader<Order> {
```
generify your reader and writer and use only one reader and writer because you are using absolutely the same approach for serialization/deserializations for all entities and I'm not sure that it will be changed in the future and for other entities

dudarenko-dmitry replied:
I agree the logic in readers and writers are similar but not the same in all of them.
They use diffenrent Mapping strategies, some of them use additional converters and switchs, one of them use BufferedReader instead of CSVReader...
Could I keep 4 different readers and writers?

Mark005 replied:
so if only mapping is different, you can separate it and pass only a mapper to read/write and provide the file name, by this way you can separate your business logic and technical implementation

Mark005 replied:
also, you can merge reader and writer in one class

Mark005 commented on src/pl/senla/hotel/ie/WriterGuest.java:
```diff
@@ -0,0 +1,32 @@
+            String[] guest = new String[]
+                    {String.valueOf(g.getIdGuest()),
+                            g.getName(),
+                            String.valueOf(g.getPhoneNumber())};
```
not null safe

dudarenko-dmitry replied:
Added checking of null inside this method. Was it really needed?
Before saving of data all Lists are checked here pl/senla/hotel/ui/ie/ExecutorImportExport.java:
    private void saveGuests() {
        if (DataStorageGuest.getDataStorageGuest().getDataList().isEmpty()) {
            System.out.println("List of Guests is empty.");
        } else {
            try {
                guestWriter.save(DataStorageGuest.getDataStorageGuest().getDataList());
            } catch (IOException e) {
                System.out.println("Error in saving of Guest: \n" + e.getMessage());
            }
        }
    }

dudarenko-dmitry replied:
added the same check to all other Writers.

Mark005 commented on src/pl/senla/hotel/ie/WriterHotelFacility.java:
```diff
@@ -0,0 +1,58 @@
+
+    private String[] convertFacilityToString(HotelFacility facility) {
+        int idFacility = facility.getIdFacility();
+        String category = facility.getCategory();
```
use Enum instead of strings

dudarenko-dmitry replied:
ready

Mark005 commented on src/pl/senla/hotel/ie/WriterHotelService.java:
```diff
@@ -0,0 +1,64 @@
+    private String[] convertHotelServiceToString(HotelService hs) {
+        String idService = String.valueOf(hs.getIdService());
+        String idOrder = String.valueOf(hs.getIdOrder());
+        String typeOfService = hs.getTypeOfService();
```
same here, Enum

dudarenko-dmitry replied:
ready for all enums in this entity

Mark005 commented on src/pl/senla/hotel/ui/ie/ImportExportMenuItem0.java:
```diff
@@ -0,0 +1,17 @@
+
+import pl.senla.hotel.ui.MenuItem;
+
+public class ImportExportMenuItem0 implements MenuItem {
```
use readable class names, if you will have 100 menu items you will need to open them one by one to find correct ones, so call it like QuitToMainMenuItem or something like that

dudarenko-dmitry replied:
done

dudarenko-dmitry committed:
TASK_6.
Change int to Integer in id-fields of Entities.

dudarenko-dmitry committed:
TASK_6.
Change addresses for serialization of Entities.

Mark005 commented on src/pl/senla/hotel/entity/CsvOrder.java:
```diff
@@ -0,0 +1,51 @@
+
+    }
+
+    public int getIdOrder() {
```
primitive

dudarenko-dmitry replied:
done

Mark005 commented on src/pl/senla/hotel/entity/CsvOrder.java:
```diff
@@ -0,0 +1,51 @@
+public class CsvOrder implements Serializable {
+
+    private Integer idOrder;
+    private int idGuest;
```
primitive

dudarenko-dmitry replied:
done

dudarenko-dmitry committed:
TASK_6.
Change String to Enum in all Facilities' entities.

dudarenko-dmitry committed:
TASK_6.
In Writers added null safe checking.

dudarenko-dmitry committed:
TASK_6.
Changed String to Enum in all HotelServices' entities.
Refactored ReaderOrder and OrderUtil.

dudarenko-dmitry committed:
TASK_6.
Renamed all "unreadable" classes in UserInterface

dudarenko-dmitry committed:
Merge remote-tracking branch 'origin/Hotel_6' into Hotel_6

Mark005 commented on src/pl/senla/hotel/entity/Guest.java:
```diff
@@ -1,11 +1,16 @@
+
+    private Integer idGuest;
     private String name;
     private int phoneNumber;
```
primitive

dudarenko-dmitry replied:
done

Mark005 commented on src/pl/senla/hotel/entity/services/HotelService.java:
```diff
@@ -5,15 +5,15 @@
+    protected Integer idService;
     protected int idOrder;
-    protected String typeOfService;
+    protected TypeOfService typeOfService;
```
just ServiceType

dudarenko-dmitry replied:
what is not correct here? String was changed to Enum.

Mark005 replied:
minor naming issue, just corrected the class name

Mark005 commented on src/pl/senla/hotel/ie/ReaderGuest.java:
```diff
@@ -0,0 +1,40 @@
+            csvToBean.setCsvReader(reader);
+            csvToBean.setMappingStrategy(getMappingStrategy());
+            return csvToBean.parse();
+        }
```
catch?

dudarenko-dmitry replied:
yes.
everything done :)
check please again

Mark005 commented:
please take a look at all primitives and think about the case when it should be null instead of the default primitive value

dudarenko-dmitry committed:
Changed all primitives to wrappers.
Added everywhere null checking of these wrappers.

dudarenko-dmitry committed:
1. Merge Reader and Writer in one ReaderWriterUniversal.
2. Injected generic in  ReaderWriterUniversal instead of old Reader and Writer for every Entity.
3. Implemented unified logic of Reading and Writing of all Entities.
4. Changed String in UserInterface to Constants.

dudarenko-dmitry committed:
deleted unused Exceptions and imports

Mark005 commented on src/pl/senla/hotel/ie/ReaderWriter.java:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    private String getPath(Class<?> clazz) {
+        if (clazz.equals(Room.class) || clazz.equals(HotelFacility.class)) {
```
it's not a good way to implement logic, my expectations were a little bit different. Better to leave in ReaderWriter logic to writing/reading a file and pass converter interface inside that will contain all methods to convert and write for specific object 

public class Converter<T> {

String getFilename();

String[] convertToString(T convertable);

T convertToObject(String[] fields);

}

dudarenko-dmitry replied:
I agree with you, Mark. 
You suggestion much better for OOP... In my implementation the Single responsibility principle doesn't work in ReaderWriter: too many actions for one class.
I could implement your comment if necessery.

Mark005 replied:
yes please, just for practice, you can not commit it if you want

Mark005 reviewed (approved):
please take a look at the comment that I left I think it will be a good practice to understand how to work with OOP and avoid casting to specific objects manually
